plutus-core/executables/debugger/Draw.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Renders the debugger in the terminal.
+module Draw where
+
+import Types
+
+import Brick.AttrMap qualified as B
+import Brick.Focus qualified as B
+import Brick.Types qualified as B
+import Brick.Widgets.Border qualified as BB
+import Brick.Widgets.Center qualified as BC
+import Brick.Widgets.Core qualified as B
+import Brick.Widgets.Edit qualified as BE
+import Data.Bifunctor
+import Data.Maybe
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Lens.Micro
+
+drawDebugger ::
+    DebuggerState ->
+    [B.Widget ResourceName]
+drawDebugger st =
+    withKeyBindingsOverlay
+        (st ^. dsKeyBindingsMode)
+        [header "Plutus Core Debugger" $ BC.center ui]
+  where
+    focusRing = st ^. dsFocusRing
+    (curLine, curCol) = BE.getCursorPosition (st ^. dsUplcEditor)
+    cursorPos = "Ln " <> show (curLine + 1) <> ", Col " <> show (curCol + 1)
+    uplcEditor =
+        BB.borderWithLabel (B.txt "UPLC program") $
+            B.vBox
+                [ B.withFocusRing
+                    focusRing
+                    (BE.renderEditor (drawDocumentWithHighlight (st ^. dsUplcHighlight)))
+                    (st ^. dsUplcEditor)
+                , B.padTop (B.Pad 1) . BC.hCenter $ B.str cursorPos
+                ]
+    sourceEditor =
+        BB.borderWithLabel (B.txt "Source program") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsSourceEditor)
+    returnValueEditor =
+        BB.borderWithLabel (B.txt "UPLC value being returned") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsReturnValueEditor)
+    cekStateEditor =
+        BB.borderWithLabel (B.txt "CEK machine state") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsCekStateEditor)
+    ui =
+        B.vBox
+            [ BC.center uplcEditor B.<+> BC.center sourceEditor
+            , B.vLimit (st ^. dsVLimitBottomEditors) $
+                BC.center returnValueEditor B.<+> BC.center cekStateEditor
+            , footer
+            ]
+
+-- | Draw a document, a consecutive part of which may be highlighted.
+drawDocumentWithHighlight ::
+    forall n.
+    -- | The part of the document to be highlighted.
+    Maybe HighlightSpan ->
+    -- | The document to draw, one `Text` per line.
+    [Text] ->
+    B.Widget n
+drawDocumentWithHighlight = \case
```

Comment:
Is there a way to make this less nested? This is a bit hard to keep track.

Reply:
You should read more pir/plc programs :joy: